### ISSUE TITLE
Rework memory allocator in order to increase contract size or call stack depth

### ISSUE THREAD

**Issue by athei** [paritytech/polkadot-sdk#6761]
Currently, the memory allocator lives in the client. It is pretty simple. For this reason we need to be very conservative how much memory we can use. This means that contract sizes and call stack depths are limited.

We should move the allocator into the runtime. This allows us to use a more sophisticated allocator.

cc @koute  AFAIK you have an allocator in mind we should use for the runtime?

- [ ] https://github.com/paritytech/polkadot-sdk/pull/7375
- [ ] https://github.com/polkadot-fellows/RFCs/pull/4
- [ ] Choose an allocator with better worst case properties so that we can reduce [`MEMORY_ALLOCATOR_INEFFICENCY_DEVIDER `](https://github.com/paritytech/polkadot-sdk/blob/472945703925a1beb094439fd7e43149c44960d5/substrate/frame/revive/src/lib.rs#L642) as much as possible
- [ ] Make `MEMORY_ALLOCATOR_INEFFICENCY_DEVIDER` configurable since it is now depending on the runtime
- [ ] Assume a certain devider and retune our contract limits to allow for deeper stacks or more code.
- [ ] Migrate the AssetHub runtimes to the newly chosen allocator.

## Alternative: Wait for the JIT

Once we move PolkaVM into the client (for the JIT) we no longer need to hold all code and memory of the current call stack in runtime memory. It will all be inside client memory and the runtime just holds a handle to it. This will likely remove the need for more memory.

My stance on this is: It will take too long to block on the JIT. We need more memory sooner to allow deeper call stacks. The linked RFC is a bit outdated (there might be new host functions it does not cover) but in good shape. It is not blocked by anything. If we get somebody to work in it, it can happen in parallel with the JIT work.

**Comment by bkchr** [comment]
This is basically: https://github.com/polkadot-fellows/RFCs/pull/4

It requires to finalize the RFC, get it approved and then we can implement it.

**Comment by athei** [comment]
Agreed. I still leave this open because I can't add the other issue to my board.

**Comment by TorstenStueber** [comment]
I removed this from H1O1KR1 as this is not required anymore for revive (quoting @athei) due to https://github.com/paritytech/polkadot-sdk/pull/9267

**Comment by athei** [comment]
As stated above we no longer require the new allocator since we found a workaround with https://github.com/paritytech/polkadot-sdk/pull/9267.

We still want to replace the host allocator eventually as it beneficial in general. It is just no longer a priority for contracts.

This work is continued as part of [RFC-145](https://github.com/polkadot-fellows/RFCs/pull/145). Closing here.
